### PR TITLE
fix: make discv4 packets adhere to eip-8

### PR DESCRIPTION
**Pull request by onbjerg**
The derived implementation of `decode` enforces that `consumed == payload_length` which does not adhere to [EIP-8](https://eips.ethereum.org/EIPS/eip-8). According to EIP-8, there may be unused list items in the packet, and these should just be skipped. 

As such, this PR replaces the derived `decode` impl with a manual one that just asserts that we did not consume *more* than `payload_length` bytes from the buffer.

### CODE REVIEW

File: crates/net/discv4/src/proto.rs [hunk 1]
```diff
@@ -215,33 +215,132 @@ impl NodeEndpoint {
 }
 
 /// A [FindNode packet](https://github.com/ethereum/devp2p/blob/master/discv4.md#findnode-packet-0x03).
-#[derive(Clone, Copy, Debug, Eq, PartialEq, RlpEncodable, RlpDecodable)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, RlpEncodable)]
 pub struct FindNode {
```
Comment (mattsse) [inline crates/net/discv4/src/proto.rs:219]: is this the same as #[rlp(trailing)]

see also:

https://github.com/paradigmxyz/reth/blob/0aa7d4d05ed5565545002655ddc94083e057977e/crates/ethereum-forks/src/forkid.rs#L125-L127

**Review by mattsse — CHANGES_REQUESTED** [review]
I see, trailing is only for optional fields -.- so we need to do this manually,

would love some sanity tests, but this looks correct

**Comment by onbjerg** [comment]
going to add some tests

**Comment by onbjerg** [comment]
Added some tests with test vectors from EIP-8.

With just `RlpDecodable` derived:

```
❯ cargo test -p reth-discv4 proto 
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.30s
     Running unittests src/lib.rs (target/debug/deps/reth_discv4-d8cf1eade260374a)

running 18 tests
test proto::tests::test_endpoint_ipv_64 ... ok
test proto::tests::decode_ping_packet ... ok
test proto::tests::eip8_decode_findnode ... FAILED
test proto::tests::eip8_decode_neighbours ... ok
test proto::tests::decode_enr_rlp ... ok
test proto::tests::encode_known_rlp_enr ... ok
test proto::tests::test_endpoint_ipv_v4 ... ok
test proto::tests::decode_pong_packet ... ok
test proto::tests::encode_decode_enr_msg ... ok
test proto::tests::neighbours_max_ipv4 ... ok
test proto::tests::test_hash_mismatch ... ok
test proto::tests::encode_decode_enr_rlp ... ok
test proto::tests::test_ping_message ... ok
test proto::tests::test_ping_message_with_enr ... ok
test proto::tests::test_pong_message ... ok
test proto::tests::test_pong_message_with_enr ... ok
test proto::tests::test_encode_decode_message ... ok
test proto::tests::neighbours_max_nodes ... ok

failures:

---- proto::tests::eip8_decode_findnode stdout ----
thread 'proto::tests::eip8_decode_findnode' panicked at crates/net/discv4/src/proto.rs:881:49:
called `Result::unwrap()` on an `Err` value: Rlp(ListLengthMismatch { expected: 78, got: 71 })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    proto::tests::eip8_decode_findnode

test result: FAILED. 17 passed; 1 failed; 0 ignored; 0 measured; 20 filtered out; finished in 1.11s

error: test failed, to rerun pass `-p reth-discv4 --lib`
```

with `#[rlp(trailing)]` as well:

```
❯ cargo test -p reth-discv4 proto
    Blocking waiting for file lock on build directory
   Compiling reth-discv4 v0.2.0-beta.6 (/Users/onbjerg/Projects/paradigmxyz/reth/crates/net/discv4)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 13.25s
     Running unittests src/lib.rs (target/debug/deps/reth_discv4-d8cf1eade260374a)

running 18 tests
test proto::tests::test_endpoint_ipv_64 ... ok
test proto::tests::neighbours_max_ipv4 ... ok
test proto::tests::decode_pong_packet ... ok
test proto::tests::decode_ping_packet ... ok
test proto::tests::eip8_decode_neighbours ... ok
test proto::tests::test_endpoint_ipv_v4 ... ok
test proto::tests::decode_enr_rlp ... ok
test proto::tests::eip8_decode_findnode ... FAILED
test proto::tests::encode_known_rlp_enr ... ok
test proto::tests::encode_decode_enr_msg ... ok
test proto::tests::test_hash_mismatch ... ok
test proto::tests::encode_decode_enr_rlp ... ok
test proto::tests::test_pong_message ... ok
test proto::tests::test_ping_message_with_enr ... ok
test proto::tests::test_ping_message ... ok
test proto::tests::test_pong_message_with_enr ... ok
test proto::tests::test_encode_decode_message ... ok
test proto::tests::neighbours_max_nodes ... ok

failures:

---- proto::tests::eip8_decode_findnode stdout ----
thread 'proto::tests::eip8_decode_findnode' panicked at crates/net/discv4/src/proto.rs:882:49:
called `Result::unwrap()` on an `Err` value: Rlp(ListLengthMismatch { expected: 78, got: 71 })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    proto::tests::eip8_decode_findnode

test result: FAILED. 17 passed; 1 failed; 0 ignored; 0 measured; 20 filtered out; finished in 1.15s

error: test failed, to rerun pass `-p reth-discv4 --lib`
```

with manual impl:

```
❯ cargo test -p reth-discv4 proto
    Blocking waiting for file lock on build directory
   Compiling reth-discv4 v0.2.0-beta.6 (/Users/onbjerg/Projects/paradigmxyz/reth/crates/net/discv4)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 22.84s
     Running unittests src/lib.rs (target/debug/deps/reth_discv4-d8cf1eade260374a)

running 18 tests
test proto::tests::test_endpoint_ipv_64 ... ok
test proto::tests::eip8_decode_findnode ... ok
test proto::tests::neighbours_max_ipv4 ... ok
test proto::tests::decode_ping_packet ... ok
test proto::tests::decode_pong_packet ... ok
test proto::tests::eip8_decode_neighbours ... ok
test proto::tests::test_endpoint_ipv_v4 ... ok
test proto::tests::encode_decode_enr_msg ... ok
test proto::tests::decode_enr_rlp ... ok
test proto::tests::encode_known_rlp_enr ... ok
test proto::tests::test_hash_mismatch ... ok
test proto::tests::encode_decode_enr_rlp ... ok
test proto::tests::test_ping_message ... ok
test proto::tests::test_pong_message ... ok
test proto::tests::test_ping_message_with_enr ... ok
test proto::tests::test_pong_message_with_enr ... ok
test proto::tests::test_encode_decode_message ... ok
test proto::tests::neighbours_max_nodes ... ok

test result: ok. 18 passed; 0 failed; 0 ignored; 0 measured; 20 filtered out; finished in 1.16s
```

**Comment by onbjerg** [comment]
Similarly with manual impl for `EnrForkIdEntry`:

```
running 1 test
test forkid::tests::eip8::eip8_decode_enr_fork_id_entry ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 8 filtered out; finished in 0.00s
```

With `#[rlp(trailing)]`:

```
failures:
    forkid::tests::eip8::eip8_decode_enr_fork_id_entry

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 8 filtered out; finished in 0.00s
```

**Review by mattsse — APPROVED** [review]

**Pull request merged**